draft-ietf-mls-protocol.md

```diff
@@ -2767,37 +2765,13 @@ def copath(x, n):
 
     return [sibling(y, n) for y in d]
 
-# Frontier is the list of full subtrees, from left to right.  A
-# balanced binary tree with n leaves has a full subtree for every
-# power of two where n has a bit set, with the largest subtrees
-# furthest to the left.  For example, a tree with 11 leaves has full
-# subtrees of size 8, 2, and 1.
-def frontier(n):
-    st = [1 << k for k in range(log2(n) + 1) if n & (1 << k) != 0]
-    st = reversed(st)
-
-    base = 0
-    f = []
-    for size in st:
-        f.append(root(size) + base)
-        base += 2*size
-    return f
-
-# Leaves are in even-numbered nodes
-def leaves(n):
-    return [2*i for i in range(n)]
-
-# The resolution of a node is the collection of non-blank
-# descendants of this node.  Here the tree is represented by a list
-# of nodes, where blank nodes are represented by None
-def resolve(tree, x, n):
-    if tree[x] != None:
-        return [x]
-
-    if level(x) == 0:
-        return []
-
-    L = resolve(tree, left(x), n)
-    R = resolve(tree, right(x, n), n)
-    return L + R
+# The common ancestor of two leaves is the lowest node that is in the
+# lowest-level node that is in the direct paths of both leaves.
+def common_ancestor(x, y):
```

Comment:
For simplicity I would just iterate over `parent()` until the value is equal.

Reply:
That actually doesn't work, because the distance to the parent might not be equal.  Consider, for example, the three-member tree:

```
    ABC
   /   \
 AB     |
 / \    |
A   B   C
```

The common ancestor of A and B is `ABC = parent(parent(A)) = parent(C)`.

You would have to first compute the direct path of one node, then iterate `parent()` on the other until you end up in the direct path.  The formula here is a little bit of index magic, but I think it can be proven correct, and it's simpler/faster.

Reply:
True, my idea only works reliably on the left part of the tree. On the right part you'd have to also use `level()` to skip some nodes. If you're confident your math works, it might just be simpler to use that.

Reply:
The algorithm passes tests on a non-trivial tree, so I have pretty good confidence that it works.  Proof sketch:

* Each subtree of the tree is identified by a common prefix
    * The root of the subtree is `<prefix>0<1 * level>`
    * The nodes within the subtree are `<prefix>0...0` to <prefix>1...10
* Therefore every common ancestor of A and B will have a prefix that is common to A and B, since a common ancestor is the root of a subtree containing both nodes
* This algorithm finds the longest prefix shared by A and B, then computes the root of the corresponding subtree

Of course, all of this is just a recommended implementation.  Implementations are free to use a more explicit algorithm like yours if they prefer.